### PR TITLE
Volunteers controller cleanup

### PR DESCRIPTION
**Pull request by rylanb**
## Overview
`def switch_user` and `def region_admin` could be cleaner and assign less variables. Plus they can use the`current_volunteer.admin_region_ids` method instead of getting it off of assignments in the controller action.

## Details
Let me know thoughts, this is code I did awhile ago and have had stashed locally. I wanted to get it up for someone to review and see if it made any sense to merge!

### CODE REVIEW

**Comment by rylanb** [comment]
@boonrs or @mooreds or @ericfreese could you run eyes past this? Super small PR, some good improvement to performance though. 

I'd prefer not to review my own code changes.

File: app/controllers/volunteers_controller.rb [hunk 1]
```diff
@@ -187,24 +183,17 @@ def super_admin
...
-      @my_admin_regions = current_volunteer.assignments.collect do |assignment|
-         assignment.admin ? assignment.region : nil
-      end.compact
-
-      admin_region_ids = @my_admin_regions.collect { |my_admin_region| my_admin_region.id }
-
       @my_admin_volunteers = Volunteer.all.collect do |volunteer|
         ((volunteer.regions.length == 0) ||
-        (admin_region_ids & volunteer.regions.collect { |region| region.id }).length > 0) ? volunteer : nil
+        (@admin_region_ids & volunteer.regions.pluck(:id)).length > 0) ? volunteer : nil
```
Comment (mooreds) [inline app/controllers/volunteers_controller.rb]: Why wouldn't you use the the join here: https://github.com/boulder-food-rescue/food-rescue-robot/blob/master/app/controllers/volunteers_controller.rb#L62

Comment (rylanb) [reply app/controllers/volunteers_controller.rb]: Great question!

**Comment by mooreds** [comment]
Looks good to me except for the two comments.

File: app/controllers/volunteers_controller.rb (hunk 1 above)
Comment (rylanb) [reply app/controllers/volunteers_controller.rb]: That was written way after I poked at this PR, I think. 

I'll have to redo now after we worked on / merged things today!

File: app/models/volunteer.rb [hunk 2]
```diff
@@ -155,12 +155,19 @@ def in_region?(region_id)
 
   def current_absences
     today = Date.today
-
     absences.where('start_date < ? AND stop_date > ?', today, today)
   end
 
+  def region_names
+    assignments.flat_map { |a| a.region.try(:name) }.uniq.join(",")
+  end
```
Comment (mooreds) [inline app/models/volunteer.rb]: Would be awesome to have a unit test for this method.

File: app/models/volunteer.rb [hunk 3]
```diff
@@ -155,12 +155,19 @@ def in_region?(region_id)
...
+
   ### CLASS METHODS
 
+  def self.with_regions_for_select(vols)
+    vols.collect { |v| ["#{v.name} ['#{v.region_names}']", v.id] }
+  end
```
Comment (mooreds) [inline app/models/volunteer.rb]: would be awesome to have a unit test for this method

File: app/controllers/volunteers_controller.rb [hunk 4]
```diff
@@ -180,21 +176,14 @@ def super_admin
...
 
   def region_admin
-    @regions = Region.all
+    @admin_region_ids = current_volunteer.admin_region_ids
+    @my_admin_regions = current_volunteer.admin_regions
+
     if current_volunteer.super_admin?
-      @my_admin_regions = @regions
-      @my_admin_volunteers = Volunteer.all
+      @my_admin_volunteers = Volunteer.includes(:regions).all
```
[the remark this answers is not in the document]
Comment (rylanb) [reply app/controllers/volunteers_controller.rb]: No this is an eager load, not a `joins`

[the remark this answers is not in the document]
Comment (mooreds) [reply app/controllers/volunteers_controller.rb]: ah, my bad

File: app/views/volunteers/region_admin.html.erb [hunk 5]
```diff
@@ -6,7 +6,7 @@
 
 <div class="row">
   <div class="col-sm-6">
-    <h2> Switch to User </h2>
+    <h2>Switch to User</h2>
     <p>
       Select a volunteer from the list below to swtich to.
```
[the remark this answers is not in the document]
Comment (rylanb) [reply app/views/volunteers/region_admin.html.erb]: Nice catch, fixed.

File: app/models/volunteer.rb (hunk 2 above)
Comment (rylanb) [reply app/models/volunteer.rb]: Updated!

File: app/models/volunteer.rb (hunk 3 above)
Comment (rylanb) [reply app/models/volunteer.rb]: Updated

File: app/views/volunteers/region_admin.html.erb [hunk 6]
```diff
@@ -18,17 +18,9 @@
...
-              <%= select_tag(:volunteer_id, options_for_select(Volunteer.find(:all, :order => 'name').collect{ |v|
-                    ["#{v.name} ["+v.regions.collect{ |r| r.name }.join(",")+"]",v.id] }), {class: "form-control select2"}) %></td>
-          <% else %>
-            <td>
-              <%= select_tag(:volunteer_id, options_for_select(Volunteer.find(:all, :order => 'name').collect{ |v|
-                    ((v.regions.collect{ |r| r.id } & @admin_region_ids).length > 0 and !v.admin) ? [v.name,v.id] : nil }.compact), {class: "form-control select2"}) %>
-            </td>
-          <% end %>
+          <td>
+            <%= select_tag(:volunteer_id, options_for_select(Volunteer.with_regions_for_select(@my_admin_volunteers)), { class: "form-control select2" }) %>
```
[the remark this answers is not in the document]
Comment (rylanb) [reply app/views/volunteers/region_admin.html.erb]: Ah ... but this issue existed previously, yes ?

[the remark this answers is not in the document]
Comment (rylanb) [reply app/views/volunteers/region_admin.html.erb]: Hmmm ... this view can only be accessed by region admins or super admins. So I'm not sure you could even get here and elevate your privileges?

[the remark this answers is not in the document]
Comment (rylanb) [reply app/views/volunteers/region_admin.html.erb]: But added here: https://github.com/boulder-food-rescue/food-rescue-robot/pull/140/commits/2a5c7697d80dacb0bb1c3bf6ed411ef4b9b8e5d0

Maybe a bit sloppy but seems to do the trick.

[the remark this answers is not in the document]
Comment (mooreds) [reply app/views/volunteers/region_admin.html.erb]: Sorry, maybe my comment wasn't clear. the v.admin! check in the old select tag isn't present in the database calls you updated. Not sure if it is important, but wanted to call it out.

[the remark this answers is not in the document]
Comment (mooreds) [reply app/views/volunteers/region_admin.html.erb]: I don't see how the commit you mentioned fixes that

File: spec/models/volunteer_spec.rb [hunk 7]
```diff
@@ -114,6 +114,62 @@
...
+
+    before do
+      create(:assignment, volunteer: volunteer_with_shifts, region: boulder)
+      create(:schedule_volunteer, volunteer: volunteer_with_shifts, schedule_chain: schedule_chain)
+      create(:assignment, volunteer: shiftless_volunteer, region: boulder)
+      create(:assignment, volunteer: shiftless_volunteer_denver, region: denver)
+    end
+
+    it 'returns region name' do
+      expect(volunteer_with_shifts.region_names).to eq('Boulder')
```
Comment (mooreds) [inline spec/models/volunteer_spec.rb]: Is it worth testing that someone in two regions gets it correctly displayed? How about someone in no regions?

Comment (rylanb) [reply spec/models/volunteer_spec.rb]: Always! Thanks for the pointer. 

How does it look to you now?

**Comment by mooreds** [comment]
Looks good to me! 👍 

Would you like me to merge (I don't know if I have permissions)?

**Pull request merged**